Write data to firestore

iritush opened the pull request:
@nfloersch @doub1ejack, this is a draft still, but if you have a moment to take a look, let me know what you think so far. I ‘started over’ with this new branch to set things up for writing to the db. I found a cveoeo file with multiple lines of actual data so I used this file to get started. 
I ended up using the System ID provided in these files as the uid. I think that if it is in fact a unique id that they use on their end, it might be clearer. In addition, it will simplify things for the (rare) case where a client decides to change the email address they want to use. We can totally change the uid to email if that is preferred.
Next step is to define the goals data and write it as a collection under the user document. I will use the sample data that Nick provided on Tuesday for what fields to have there

doub1ejack commented on functions/index.js:
```diff
@@ -133,28 +154,88 @@ exports.pullDataFromSftp= functions.https.onRequest((request, response) => {
+  usersCollection.where('uid', '==', user.uid).get()
+    .then(snapshot => {
+      if (snapshot.empty) {
+        console.log('No matching documents. Creating a new document for this user');
```
```suggestion
        console.log('No matching documents. Creating a new document for this user');
          // TODO: create a validate function on the user model and use it to ensure that we only
          // push valid user info to Firestore. I'm not sure I trust Outcome Tracker's sftp dumps enough 
          // to assume that they will always be giving us all the info we need
```
I like making a habit of leaving `// TODO`s in the code, and creating corresponding issues in the github project when I think of something tangental to the current work.  Validating the user isn't a vital part of this PR, so it's a good 'todo' for later.  You _could_ do it as part of this PR, but keeping PRs and issues small makes it more likely that each one will be completed quickly, and also allows other developers to contribute little pieces too.

doub1ejack commented on functions/index.js:
```diff
@@ -15,6 +21,21 @@ exports.helloWorld = functions.https.onRequest((request, response) => {
 	 }
 });
 
+exports.pullDataFromLocalCSVFile = functions.https.onRequest((request, response) => {
```
I don't see where this function is used.  If it is something you're using for testing purposes (and not intended to be part of the working system), would you clarify that in a comment at the beginning of the function?  Thanks 🙏

iritush replied:
Yes. This is to be able to test with a file saved locally. I will add a comment

iritush commented on functions/index.js:
```diff
@@ -133,28 +154,88 @@ exports.pullDataFromSftp= functions.https.onRequest((request, response) => {
+          displayName: user.firstName,
+          lastName: user.lastName,
+          email: user.email
+        };
```
Makes sense. I will take a look

iritush replied:
Added a TODO item for this

iritush commented on functions/models/validators.js:
```diff
@@ -0,0 +1,4 @@
+const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
I will take a look at the existing library you referenced here. (I actually copied this regex from libs/validators.js :))

iritush replied:
Added a TODO item for this

iritush commented:
In this branch:
-Removed updateOrCreateUseronFireStore(). Instead, write data to the Firestore database using class methods in the ‘user’ class
-added a 'goal' class
-Changed some function and variable names for clarity
-Added comments, TODOs, and did some cleanup
-Known issue: When parsing a csv file with multiple lines that have goal data, saving to firestore is not working properly (working on it :-))
TODOs to point out: 
-look into tracking the name of the last parsed file and pulling all new files that were 
  added to the server since then
-confirm with cvoeo that the 'System Name ID' field is a reliable unique id to use

doub1ejack commented on functions/models/user.js:
```diff
@@ -0,0 +1,54 @@
+        "Last name: " + this.lastName + "\n" +
+        "email: " + this.email + "\n")
+    }
+    createNewUserInFirestore() {
```
I think this work probably addresses task #157 on the [project board](https://github.com/codeforbtv/cvoeo-app/projects/1).  Would you review that issue and either update it to reflect any changes you feel still need to be made, or just tag it with this PR number (169) and move it to the completed column?
